File: sdk/py/src/obsigna/emitters/buffering.py
"""BufferingEmitter — in-memory batch buffer with timer flush (ADR-0020).

Wraps a downstream :class:`Emitter` and buffers receipts in memory,
flushing on a configurable interval or batch size.

The contract with the downstream emitter is per-receipt, NOT batched:
a flush calls ``inner.emit(receipt)`` once per buffered receipt.

!!! CRASH-LOSS RISK !!!
Buffered receipts are lost if the process exits before :meth:`flush`
completes. This emitter is NOT suitable for environments where audit
completeness is critical. Use a synchronous :class:`HttpEmitter` (or a
persistent WAL — tracked separately) when every receipt must reach the
collector.
"""

from __future__ import annotations

import logging
import threading
from typing import TYPE_CHECKING

from obsigna.emitters.types import BufferingFlushError

if TYPE_CHECKING:
    from obsigna.emitters.types import Emitter
    from obsigna.receipt.types import AgentReceipt


logger = logging.getLogger(__name__)


class BufferingEmitter:
    """Per-receipt batching wrapper around a downstream :class:`Emitter`.

    Parameters
    ----------
    inner:
        Downstream emitter. Each buffered receipt is delivered with a
        separate ``inner.emit(receipt)`` call when the buffer flushes.
    max_batch_size:
        Flush when the buffer reaches this many receipts. Must be >= 1.
    flush_interval_ms:
        Flush every N milliseconds while receipts are buffered. Must be
        >= 1. The interval timer only runs when at least one receipt is
        buffered — there is no idle-process tick.
    """

    def __init__(
        self,
        *,
        inner: Emitter,
        max_batch_size: int,
        flush_interval_ms: int,
    ) -> None:
        if max_batch_size < 1:
            raise ValueError("BufferingEmitter: max_batch_size must be >= 1")
        if flush_interval_ms < 1:
            raise ValueError("BufferingEmitter: flush_interval_ms must be >= 1")
        self._inner = inner
        self._max_batch_size = max_batch_size
        self._flush_interval = flush_interval_ms / 1000.0

        self._lock = threading.Lock()
        # Held across every delivery loop so concurrent emit() callers
        # (or an emit racing with the timer) cannot interleave batches at
        # the inner emitter. The two locks must NEVER be held together:
        # always release self._lock before acquiring self._delivery_lock.
        self._delivery_lock = threading.Lock()
        self._buffer: list[AgentReceipt] = []
        self._timer: threading.Timer | None = None
        self._closed = False

    def emit(self, receipt: AgentReceipt) -> None:
        with self._lock:
            if self._closed:
                raise RuntimeError("BufferingEmitter: closed")
            self._buffer.append(receipt)
            if len(self._buffer) >= self._max_batch_size:
                # Drop the buffer lock before flushing — downstream
                # emit() may take a long time and we don't want
                # concurrent emit() callers to stall on it. Ordering is
                # preserved by self._delivery_lock around _deliver.
                batch = self._buffer[:]
                self._buffer.clear()
                self._cancel_timer_locked()
            else:
                self._schedule_timer_locked()
                return
        self._deliver(batch)

    def flush(self) -> None:
        """Drain the buffer through the downstream emitter.

        If one or more downstream calls raise, the rest of the batch is
        still attempted and a :class:`BufferingFlushError` carrying every
        failure is raised at the end. Retries are the inner emitter's
        responsibility — failed receipts are not re-queued.
        """
        with self._lock:
            self._cancel_timer_locked()
            batch = self._buffer[:]
            self._buffer.clear()
        self._deliver(batch)

    def close(self) -> None:
        """Stop the interval timer and flush the remaining buffer."""
        with self._lock:
            if self._closed:
                return
            self._closed = True
            self._cancel_timer_locked()
            batch = self._buffer[:]
            self._buffer.clear()
        self._deliver(batch)

    # ------------------------------------------------------------------

    def _deliver(self, batch: list[AgentReceipt]) -> None:
        # Per-receipt delivery serialised through self._delivery_lock so
        # concurrent emits / explicit flush / timer flush never
        # interleave batches at the inner emitter.
        if not batch:
            return
        with self._delivery_lock:
            errors: list[BaseException] = []
            for receipt in batch:
                try:
                    self._inner.emit(receipt)
                except (
                    Exception  # noqa: BLE001 — aggregate; surface every failure
                ) as exc:
                    errors.append(exc)
        if not errors:
            return
        if len(errors) == 1:
            raise errors[0]
        raise BufferingFlushError(
            f"BufferingEmitter: {len(errors)} of {len(batch)} receipts failed",
            errors,
        )

    def _schedule_timer_locked(self) -> None:
        # Caller must hold self._lock. Idempotent: a running timer is
        # not replaced — emit() relies on the existing tick.
        if self._timer is not None:
            return
        t = threading.Timer(self._flush_interval, self._on_timer)
        t.daemon = True
        self._timer = t
        t.start()

    def _cancel_timer_locked(self) -> None:
        if self._timer is not None:
            self._timer.cancel()
            self._timer = None

    def _on_timer(self) -> None:
        with self._lock:
            self._timer = None
            if self._closed:
                return
            batch = self._buffer[:]
            self._buffer.clear()
        if not batch:
            return
        # Log at debug — the timer thread must not crash on downstream
        # errors. Callers that need surfaced errors should rely on the
        # explicit flush()/close() path which raises.
        try:
            self._deliver(batch)
        except Exception as exc:  # noqa: BLE001 — timer must not crash
            logger.debug(
                "BufferingEmitter: timer flush failed",
                extra={"err": str(exc)},
            )

File: sdk/go/emitters/buffering.go
package emitters

import (
	"context"
	"errors"
	"log/slog"
	"sync"
	"time"

	"obsigna.dev/sdk/go/receipt"
)

// BufferingEmitterConfig configures [BufferingEmitter].
type BufferingEmitterConfig struct {
	// Inner is the downstream emitter that receives per-receipt deliveries
	// when the buffer flushes. Required.
	Inner Emitter

	// MaxBatchSize triggers a flush as soon as the buffer reaches this
	// size. Must be >= 1.
	MaxBatchSize int

	// FlushInterval triggers a flush after this duration of buffered
	// inactivity. Must be > 0.
	FlushInterval time.Duration

	// Logger is used for timer-path delivery failures. Defaults to
	// slog.Default(); pass a discard handler if you want silence.
	Logger *slog.Logger
}

// BufferingEmitter wraps a downstream [Emitter] and buffers receipts in
// memory, flushing them on a configurable interval or batch size.
//
// The contract with the downstream emitter is PER-RECEIPT, not batched:
// a flush calls Inner.Emit once per buffered receipt.
//
// !!! CRASH-LOSS RISK !!!
// Buffered receipts are lost if the process exits before [BufferingEmitter.Flush]
// completes. This emitter is NOT suitable for environments where audit
// completeness is critical. Use a synchronous [HttpEmitter] (or a
// persistent WAL — tracked separately) when every receipt must reach the
// collector.
type BufferingEmitter struct {
	inner         Emitter
	maxBatchSize  int
	flushInterval time.Duration
	logger        *slog.Logger

	// mu guards the buffer + timer + closed fields.
	mu     sync.Mutex
	buffer []receipt.AgentReceipt
	timer  *time.Timer
	closed bool

	// deliveryMu serialises every call into the inner emitter so
	// concurrent Emit, Flush, Close, and timer-driven flushes can never
	// interleave batches at the downstream — required for hash-chained
	// audit trails. Acquire AFTER releasing mu; never hold both at once.
	deliveryMu sync.Mutex
}

// NewBuffering constructs a [BufferingEmitter] from the given config.
// Returns an error for invalid sizes/intervals so a misconfiguration is
// caught at construction rather than on the first Emit.
func NewBuffering(cfg BufferingEmitterConfig) (*BufferingEmitter, error) {
	if cfg.Inner == nil {
		return nil, errors.New("BufferingEmitter: Inner is required")
	}
	if cfg.MaxBatchSize < 1 {
		return nil, errors.New("BufferingEmitter: MaxBatchSize must be >= 1")
	}
	if cfg.FlushInterval <= 0 {
		return nil, errors.New("BufferingEmitter: FlushInterval must be > 0")
	}
	logger := cfg.Logger
	if logger == nil {
		logger = slog.Default()
	}
	return &BufferingEmitter{
		inner:         cfg.Inner,
		maxBatchSize:  cfg.MaxBatchSize,
		flushInterval: cfg.FlushInterval,
		logger:        logger,
	}, nil
}

// Emit appends r to the buffer and either flushes immediately (when
// MaxBatchSize is reached) or schedules a flush after FlushInterval.
func (b *BufferingEmitter) Emit(ctx context.Context, r receipt.AgentReceipt) error {
	b.mu.Lock()
	if b.closed {
		b.mu.Unlock()
		return errors.New("BufferingEmitter: closed")
	}
	b.buffer = append(b.buffer, r)
	if len(b.buffer) >= b.maxBatchSize {
		batch := b.takeBatchLocked()
		b.mu.Unlock()
		return b.deliver(ctx, batch)
	}
	b.scheduleTimerLocked()
	b.mu.Unlock()
	return nil
}

// Flush delivers every buffered receipt through the inner emitter.
// Resolves once each one has been delivered (or the inner emitter
// returned an error for one of them).
func (b *BufferingEmitter) Flush(ctx context.Context) error {
	b.mu.Lock()
	batch := b.takeBatchLocked()
	b.mu.Unlock()
	return b.deliver(ctx, batch)
}

// Close stops the interval timer and flushes the remaining buffer. After
// Close, subsequent Emit calls return an error. Safe to call multiple
// times.
func (b *BufferingEmitter) Close(ctx context.Context) error {
	b.mu.Lock()
	if b.closed {
		b.mu.Unlock()
		return nil
	}
	b.closed = true
	batch := b.takeBatchLocked()
	b.mu.Unlock()
	return b.deliver(ctx, batch)
}

// ----------------------------------------------------------------------

// takeBatchLocked drains the buffer and cancels any pending timer.
// Caller must hold b.mu.
func (b *BufferingEmitter) takeBatchLocked() []receipt.AgentReceipt {
	if b.timer != nil {
		b.timer.Stop()
		b.timer = nil
	}
	if len(b.buffer) == 0 {
		return nil
	}
	batch := b.buffer
	b.buffer = nil
	return batch
}

func (b *BufferingEmitter) scheduleTimerLocked() {
	if b.timer != nil {
		return
	}
	b.timer = time.AfterFunc(b.flushInterval, b.onTimer)
}

func (b *BufferingEmitter) onTimer() {
	b.mu.Lock()
	b.timer = nil
	if b.closed {
		b.mu.Unlock()
		return
	}
	batch := b.buffer
	b.buffer = nil
	b.mu.Unlock()
	if len(batch) == 0 {
		return
	}
	// Log timer-path failures at debug — the timer goroutine must never
	// crash the host program on a downstream failure. Callers should
	// rely on Flush / Close for error surfacing.
	if err := b.deliver(context.Background(), batch); err != nil {
		b.logger.Debug("BufferingEmitter: timer flush failed",
			slog.String("err", err.Error()),
		)
	}
}

// deliver pushes every receipt in batch through the inner emitter,
// serialised by deliveryMu so no two delivery loops can interleave at
// the downstream. Errors from individual receipts are accumulated and
// returned as one errors.Join — every receipt is attempted even when
// earlier ones fail, so the tail of a batch is never silently dropped
// (retries are the inner emitter's responsibility, not ours).
func (b *BufferingEmitter) deliver(ctx context.Context, batch []receipt.AgentReceipt) error {
	if len(batch) == 0 {
		return nil
	}
	b.deliveryMu.Lock()
	defer b.deliveryMu.Unlock()
	var errs []error
	for _, r := range batch {
		if err := b.inner.Emit(ctx, r); err != nil {
			errs = append(errs, err)
		}
	}
	return errors.Join(errs...)
}
